### DISCUSSION TITLE
Return `useTransactor`'s return value in `useScaffoldContractWrite`'s `writeAsync` function

### DISCUSSION THREAD

**Discussion by Pernick** [scaffold-eth/scaffold-eth-2#585] [Q&A] [answered]
Hello, I am wondering about [this line](https://github.com/scaffold-eth/scaffold-eth-2/blob/ab6c9c3decceb961657fc9ddbdf7e0b56d6eb324/packages/nextjs/hooks/scaffold-eth/useScaffoldContractWrite.ts#L71) in `useScaffoldContractWrite` hook, specifically why it doesn't return the value from `writeTx` function.

```
return await writeTx( ... );
// instead of
await writeTx( ... );
```

**My use case:**
I am using the `writeAsync` function returned from the `useScaffoldContractWrite` hook in form submit event. It seems like I could use the return value (`'0x${string}' | undefined`) to determine whether the write was successful (and reset the form for example).

**My questions:**
1. Is there any reason not to return the value in the `writeAsync` function?
2. Is it safe to assume that when `undefined` returns the write was not successful?
3. Am I missing some obviously superior approach?

Thanks!

**Comment by technophile-04** [comment, marked as answer]
Just to be sure you can rely on  [`onSucess`](https://wagmi.sh/react/hooks/useContractWrite#onsettled-optional) which runs after the write is successful : 

```ts
const { isSuccess, data } = useScaffoldContractWrite({
  contractName: "YourContract",
  functionName: "setGreeting",
  args: ["Hello World"],
  onSuccess(data) {
    console.log(data.hash);
    resetForm();
  },
});
```

Alternatively, you could also rely on `isSucess` variable. 

>  It seems like I could use the return value ('0x${string}' | undefined) 

The return value `('0x${string}' | undefined)` is actually the transaction `hash` which you can access through `data` property returned by `useScaffoldContractWrite` or even `onSucess` function data.hash gives you same value 🙌

**Reply by Pernick** [reply]
Thanks for the reply!

> Just to be sure you can rely on [onSucess](https://wagmi.sh/react/hooks/useContractWrite#onsettled-optional) which runs after the write is successful

I am aware of this, what I was wondering is whether I could return the `hash` from `useScaffoldContractWrite` interchangeably with the `onSuccess` callback - meaning if the `hash` returns its the same as if the `onSuccess` would have been called and there are no exceptions that I am not seeing now.
This would allow for different component composition (e.g. `useScaffoldContractWrite` not being in the form component).

But there are probably other solutions to this, like creating a store with reducer function and have these callbacks dispatch actions and just pass the store to the form.

Thanks!
